### PR TITLE
compression: enable use of decompression errors

### PR DESCRIPTION
**Pull request by derekargueta**
Description: The decompression library uses a `RELEASE_ASSERT` to verify there are no decompression errors. This is because today the decompression library is only used for testing the gzip compression filter, however this makes it impractical for anyone to use the decompression library for a filter.
Risk Level: Low
Testing: Added
Docs Changes: N/A
Release Notes: N/A

cc @gsagula 

Signed-off-by: Derek Argueta <dereka@pinterest.com>

### CODE REVIEW

**Comment by jmarantz** [comment]
Should https://github.com/envoyproxy/envoy/pull/7057 be merged first?

**Comment by derekargueta** [comment]
@jmarantz I don't think it actually matters which lands first - having looked through that PR it doesn't touch the decompression code at all, just compression. The decompressor is currently only used by testing code to validate the output of the gzip filter:
```
$ git grep "ZlibDecompressorImpl" | cut -d ':' -f 1 | uniq
source/common/decompressor/zlib_decompressor_impl.cc
source/common/decompressor/zlib_decompressor_impl.h
test/common/compressor/compressor_fuzz_test.cc
test/common/decompressor/zlib_decompressor_impl_test.cc
test/extensions/filters/http/gzip/gzip_filter_integration_test.cc
test/extensions/filters/http/gzip/gzip_filter_test.cc
```

but I'm perfectly happy to wait for the gzip generalization PR to land, no rush on this.

File: include/envoy/decompressor/decompressor.h [hunk 1]
```diff
@@ -16,9 +16,10 @@ class Decompressor {
    * Decompresses data from one buffer into another buffer.
    * @param input_buffer supplies the buffer with compressed data.
    * @param output_buffer supplies the buffer to output decompressed data.
+   * @return the status code from decompression. A negative integer indicates failure.
```
Comment (jmarantz) [inline include/envoy/decompressor/decompressor.h]: I'd suggest either returning a richer error status of some sort, either an enum described in this interface (ie not Z_OK et al) or bool+string for an expanded error message. Or you can just return a bool.

Comment (derekargueta) [reply include/envoy/decompressor/decompressor.h]: I like the idea of mapping an enum to the library return codes.

Comment (jmarantz) [reply include/envoy/decompressor/decompressor.h]: @derekargueta  are you going to make a commit to do that?

**Comment by derekargueta** [comment]
@jmarantz going to revisit this later, it hasn't been high in my queue. I'm also thinking it'd be nice to have `compress` and `decompress` have a symmetrical API of returning an error code - it'd be odd for only decompress to return the error to the user.

**Comment by jmarantz** [comment]
Sure, symmetry is a good goal. But what kind of error would you get while
compressing, other than maybe "out of memory"?

**Comment by derekargueta** [comment]
hmm that's a great point. Compression tends to be error-free (besides system issues i.e. memory), whereas decompression mandates a known format that, if corrupt or invalid in any other way, can easily surface an actual error that should be handled by the user. From that perspective, the asymmetry actually seems fine.

**Pull request closed without merging**